### PR TITLE
Feature/kernel stein discrepancy

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Tests

### Description

Adds Kernelised Stein Discrepancy.
See Exercise 3 from https://arxiv.org/pdf/2109.06075 for analytic test.
See definition of $L\text{-}KSD_{\lambda}^2$ in https://arxiv.org/pdf/2301.13528 (pg 6) for version of KSD implemented 

### How Has This Been Tested?

Adds unit tests for KSD.

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: tests/unit/test_metrics.py [hunk 1]
```diff
@@ -216,3 +222,165 @@ def test_mmd_random_data(
...
+
+        base_kernel = SquaredExponentialKernel()
+        kernel = convert_stein_kernel(x.data, base_kernel, None)
+        metric = KSD(kernel=kernel, score_matching=None)
+
+        # Compute each term in the KSD formula to obtain an expected KSD.
+        kernel_mm = kernel.compute(y.data, y.data)
+        if mode == "weighted":
+            weights_mm = y.weights[..., None] * y.weights[None, ...]
+            expected_ksd = jnp.average(kernel_mm, weights=weights_mm)
```
Comment (pc532627) [inline tests/unit/test_metrics.py:354]: Is there a reason we use `jnp.average` here but `jnp.mean` below?

**Review by pc532627 — CHANGES_REQUESTED** [review]
A few questions about a factor of `1/m` in a couple of places - otherwise only very minor points

File: coreax/metrics.py [hunk 2]
```diff
@@ -131,3 +134,133 @@ def compute(
...
+       :class:`~coreax.score_matching.KernelDensityMatching` unless 'kernel' is a
+       :class:`~coreax.kernel.SteinKernel`, in which case the kernel's existing score
+       function is used.
+    :param precision_threshold: Threshold above which negative values of the squared KSD
+        are rounded to zero (accommodates precision loss)
+    """
+
+    kernel: coreax.kernel.Kernel
+    score_matching: Optional[ScoreMatching] = None
+    precision_threshold: float = 1e-12
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:191]: Fixed

File: coreax/metrics.py [hunk 3]
```diff
@@ -131,3 +134,133 @@ def compute(
...
+        :return: Kernel Stein Discrepancy as a 0-dimensional array
+        """
+        del kwargs
+        # Train Stein kernel with data from P
+        x, w_x = jtu.tree_leaves(reference_data)
+        kernel = convert_stein_kernel(x, self.kernel, self.score_matching)
+
+        # Variable rename allows for nicer automatic formatting.
+        y = comparison_data
+        squared_ksd = kernel.compute_mean(y, y, block_size=block_size, unroll=unroll)
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/metrics.py:243]: The code in `kernel.compute_mean` actually does divide by `1/len(y)^2` (see line 319 of `kernel.py`) so it is consistent

File: tests/unit/test_metrics.py [hunk 4]
```diff
@@ -216,3 +222,165 @@ def test_mmd_random_data(
...
+        def _stein_kernel(x: ArrayLike, y: ArrayLike) -> Array:
+            norm_sq = jnp.linalg.norm(jnp.subtract(x, y)) ** 2
+            body = 1 + norm_sq
+            first_term = 3 * norm_sq / body ** (5 / 2)
+            second_term = (dim - norm_sq) / body ** (3 / 2)
+            third_term = jnp.dot(x, y) / body ** (1 / 2)
+            return -first_term + second_term + third_term
+
+        stein_kernel_matrix = pairwise(_stein_kernel)(comparison_pts, comparison_pts)
+        expected_output = stein_kernel_matrix.sum() / num_data_points**2
```
[the remark this answers is not in the document]
Comment (db091756) [reply tests/unit/test_metrics.py:304]: See above explanation for `kernel.compute_mean`

File: tests/unit/test_metrics.py (hunk 1 above)
Comment (db091756) [reply tests/unit/test_metrics.py:354]: I use `jnp.average` in order to make correct use of the weights, below we do an unweighted mean so no need for it.

**Review by pc532627 — APPROVED** [review]

**Pull request merged**